use pandas for domain values load and update SF-133 loader

bsweger opened the pull request:
Wanted to show the changes in progress to our domain load process. In particular, wanted to get @jworcestBAH's feedback on the work is divided between `loadFile.py` and `loaderUtils.py`.

The basic thought process was...for each set of domain values:
* delete existing database rows, if necessary (this varies depending on what's being loaded)
* read the .csv
* call the common data cleaning function
* do any post-processing specific that particular set of domain values (e.g., tas field in SF-133)
* call common db insert routine function

jworcestBAH commented:
The approach seems fine but I'm wondering if we could move the csv loading and deletion of previous records into LoaderUtils, those seems to be pretty similar for each loadFile function.

bsweger commented:
@jworcestBAH thanks for taking a look--I moved that function to `loaderUtils.py`. I also re-instated the `strip_commas` option (which I'd missed) and removed the `skip_duplicates` option. 

The logic for that latter decision is that how do de-dupe a file is specific to the file, and because sometimes we need to de-dupe on multiple columns, squeezing that into the current options structure became awkward.

bsweger commented:
This is still a WIP because there will be conflicts once the unittest branch is merged. In addition, I need to add a check to zero out specific SF 133 line numbers if necessary (still have to look at the data to see if we need to do that).

That said, what's here so far is ready for review. There's substantial changes to how we're loading domain values, so if folks can review this now, reviewing the remaining work should be a lot faster.

cmc333333 commented on dataactvalidator/filestreaming/csvS3Reader.py:
```diff
@@ -24,18 +18,23 @@ def openFile(self,region,bucket,filename,csvSchema,bucketName,errorFilename):
         self.isLocal = False
-        if(self.s3File == None):
-            raise ValueError("".join(["Filename provided not found on S3: ",str(filename)]))
+        if not self.s3File:
```
Very minor, but recommended practice for testing against `None` (as opposed to `False`, the empty string, etc) is:
```python
if self.s3File is None:
```

cmc333333 commented on dataactvalidator/filestreaming/csvS3Reader.py:
```diff
@@ -24,18 +18,23 @@ def openFile(self,region,bucket,filename,csvSchema,bucketName,errorFilename):
-            raise ValueError("".join(["Filename provided not found on S3: ",str(filename)]))
+        if not self.s3File:
+            raise ValueError("Filename {} not found in bucket {}".format(
+                filename, bucket))
```
This function is almost identical to `openFile`; would it make sense for one to call the other or to pull out the shared code into a `initializeS3File` method?

cmc333333 commented on dataactvalidator/filestreaming/csvS3Reader.py:
```diff
@@ -53,9 +52,10 @@ def _getNextPacket(self):
         offsetCheck = self.packetCounter *  CsvAbstractReader.BUFFER_SIZE 
-        header ={'Range' : "".join(['bytes=',str(offsetCheck),'-',str(offsetCheck +CsvAbstractReader.BUFFER_SIZE - 1)])}
+        header ={'Range': "".join(
+            ['bytes=', str(offsetCheck), '-', str(offsetCheck +CsvAbstractReader.BUFFER_SIZE - 1)])}
```
Super minor, but this is another candidate for `str.format()`

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+    # read CGAC values from csv
+    data = pd.read_csv(filename, dtype=str)
+    # toss out rows with missing CGAC codes
+    data = data[data['CGAC'].notnull()]
```
Slick!

bsweger replied:
As it turns out, I added this to handle a missing CGAC code value in our .csv. That missing value should actually be `000`, so ended up removing this line.

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+        return data
+
+    @classmethod
+    def insertDataframe(cls, df, table, engine):
```
Hmmm. I'm torn here. I see benefit in allowing `pandas` to do its thing wrt SQL generation, but I'm concerned about adding a non-ORM route to manipulating the data.

Is there a place to put this that'd fit into the ORM a bit better? Maybe within the ORM models or on the relevant "interface" object? This might also allow you to add a validation that the right columns are present.

bsweger replied:
I'm torn too. We had discussed the ramifications of bypassing the ORM shortly before joined us, and agreed that we could do so as part of the domain values load.

That said, this did bite me a few times during development and testing, so I'm amenable to re-thinking this. However, I'm thinking that the benefits of getting this in sooner so we can test the pieces of the load and making sure the deploy works with all this new data outweigh the costs of refactoring at this point.

What do you think?

cmc333333 replied:
Sounds good. To be super clear, these comments are meant to be suggestions only. I think the only thing that's blocking this PR is tests at this point.

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+
+    data = pd.read_csv(filename, dtype=str)
+    # toss out blank rows
+    data.dropna(inplace=True)
```
It might be nice to document why we're using `dropna` here, `notnull` elsewhere, and neither in other places.

bsweger replied:
Good point. This `dropna` (oof, which should actually be using the `how=all` parameter) is intended to handle the case where an incoming .csv has a bunch of extraneous blank rows at the end. That's universal enough that I moved (the corrected version of) this line to the `cleanData` function in `loaderUtils.py`.

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+    model = SF133
+
+    existing_records = interface.session.query(model).filter(
+        and_(model.fiscal_year == fiscal_year, model.period == fiscal_period))
```
Curious: do you need the `and_` here? It looks like it's implicit: http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.filter

bsweger replied:
You're right--thanks..will take that out.

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+        and_(model.fiscal_year == fiscal_year, model.period == fiscal_period))
+    if force_load:
+        # force a reload of this period's current data
+        print('Force SF 133 load: deleting existing records for {} {}'.format(
```
In general, `logging` is better than `print`s; it allows more configuration, storage/redirection, etc. Can we start that trend here? Something like:

```python
import logging
# ...
logger = logging.getLogger(__name__)
# ...
logger.info('Force SF 133 load: deleting existing records for %s %s',
            fiscal_year, fiscal_period)
```

bsweger replied:
Yes, great idea. There are several functions in loadFile.py that print, and I got the logging working by setting up the logger at the beginning of each.

What's the recommended approach to avoid this repetitive code? Put all this stuff in a class? Decorator? Global variable of some kind?

cmc333333 replied:
The strategy I learned under @vrajmohan and @jmcarp's excellent tutelage is to create a module-level `logger` and configure everything at the app's entry points. It'd look something like this:

```python
# my/module/here.py
import logging

logger = logging.getLogger(__name_)

def some_func():
    logger.info("Some Func called")    # or debug, warning, etc.

# run.py
import logging

if __name__ == '__main_':
    logging.basicConfig()    # Can be configured with threshold level here
    runProgram()
```

A better long-term strategy would be to include config information in the project's settings so it can be configured without modifying code -- Django does this, but Flask's intentionally minimal.

bsweger replied:
If @vrajmohan and @jmcarp recommend this approach, then it shall be so. Thanks!

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+    # line numbers 2002 and 2012 are the only duped line numbers,
+    # and they are not used by the validation rules, so for now
+    # just remove them before loading our SF-133 table
+    dupe_line_numbers = ['2002', '2102']
```
Thanks for the comments! I don't understand them yet, but they're much better than having magic numbers

bsweger replied:
I reworded the comment slightly to make it clearer that the line numbers in question are not surrogate keys or line numbers in a file...they represent specific designations on a budget report that really shouldn't be duplicated (however, we do not validate against those lines, hence the skipping until we dig a little deeper).

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+    data = data[~data.line.isin(dupe_line_numbers)]
+
+    # add concatenated TAS field for internal use (i.e., joining to staging tables)
+    data['tas'] = data.apply(lambda row: formatInternalTas(row), axis=1)
```
Also slick!

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+def formatInternalTas(row):
+    """Concatenate TAS components into a single field for internal use."""
+    # This formatting should match formatting in dataactcore.models.stagingModels concatTas
+    tas = '{}{}{}{}{}{}{}'.format(
```
This is a case where the `''.join` approach might actually be easier to read

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
     # SF 133 is kept on S3, so need to download that
     reader = CsvS3Reader()
 
     if localSFPath is not None:
```
There's a lot of new logic between these two branches of the `if` -- would it make sense to chunk them out into separate methods/functions?

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+                    os.linesep, sf133))
+                continue
+            year = file.split('_')[-2]
+            period = file.split('_')[-1]
```
This text munging doesn't feel good. Would a regex be easier to understand?

```python
SF_RE = re.compile(r'sf_133_(?P<year>\d{4})_(?P<period>\d+)\.csv')
# ... 
match = SF_RE.match(sf133)
if not match:
    # log error
    continue
loadSF133(sf133, match.group('year'), match.group('period'))
```

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
+                year = file.split('_')[-2]
+                period = file.split('_')[-1]
+                print('{}Starting {}...'.format(os.linesep,sf133.name))
+                loadSF133(sf133, year, period)
```
This is _very_ similar to the `localSFPath is not None` path above. Can they be combined?

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
-            interface: interface to DB table is in
-            fieldMap: dict that maps columns of the csv to attributes of the ORM object
+            data = dataframe of domain values
+            fieldMap: dict that maps columns of the dataframe csv to our db columns
```
I recognize that the `fieldMap` wasn't introduced in this PR, but it feels hairy. Is there a way to define those values statically instead of at method invocation time?

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+                 "pad_to_length" which if present will pad the field with leading zeros up to
+                specified length
+                "skip_duplicate" which ignores subsequent lines that repeat values
+                "strip_commas" which removes commas
```
Again, I know these options weren't introduced here, but this sounds like a great use case for a `namedtuple` rather than a dict

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+        # Fix up cells that have spaces instead of being empty.
+        # Set the truly empty cells to None so they get inserted to db as NULL
+        # TODO: very ugly function below...is there a better way?
+        data.applymap(lambda x: str(x).strip() if len(str(x).strip()) else None)
```
Maybe two steps?
```python
data.applymap(lambda x: str(x).strip())
data.applymap(lambda x: x or None)
```
Pandas might have some helpful function, too?

bsweger replied:
Thanks for taking a look at this. The applymap is pretty expensive to do on these large dataframes, so I'm hesitant to do it twice. I've been through the Pandas stuff and couldn't find anything magic that will handle the case where a cell has spaces. I did play with setting whitespace cells to NaN when the file was read, but that cascading into some other issues.

In conclusion--it would be good to make this process less fragile but am hesitant to pull this thread.

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -1,10 +1,13 @@
+    # pandas dataframes as necessary
+    padFunction = lambda field, padTo: str(field).strip().zfill(padTo)
+    currentTimeFunction = lambda x: datetime.utcnow()
+    cleanColNamesFunction = lambda field: str(field).lower().strip().replace(" ","_").replace(",","_")
```
Any reason these are static variables as opposed to straight functions? They don't seem too tightly coupled with `LoaderUtils` and a `def` is generally easier to read than a `lambda`

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+        # make sure all values in fieldMap parameter are in the dataframe/csv file
+        for field in fieldMap:
+            if field not in list(data.columns):
+                raise ValueError("{} is required for loading table{}".format(field, model))
```
Noting that this is a prime example for why the `fieldMap` approach is problematic. Rather than relying of methods/functions with named arguments to verify the parameters at call-time (and potentially before, using static analysis), we're throwing an exception half way through this function.

bsweger replied:
So, namedtuples seem pretty cool. What I'm not sure of is how, in this scenario, they would help us catch this error at call time.

The thing I don't see yet is how switching the `cleanData` params from dictionaries to named tuples will prevent having to do this check, which is making sure that the passed items are supplied as headers in the incoming .csv (which need to be cleaned up a bit before the check).

Can you elaborate?

cmc333333 replied:
Ah, I misread this as validating `fieldMap` rather than validating `data.columns`. Sorry for the confusion!

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+        for col, options in fieldOptions.items():
+            if "pad_to_length" in options:
+                # pad to specified length
+                data['{}'.format(col)] = data['{}'.format(col)].apply(
```
Can these indexes be shorted to `data[col]` or `data[str(col)]`?

bsweger replied:
Yes, yes they can.

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+            if "pad_to_length" in options:
+                # pad to specified length
+                data['{}'.format(col)] = data['{}'.format(col)].apply(
+                    lambda x: cls.padFunction(x, options['pad_to_length']))
```
NBD, but if we're passing around lots of functions and lambdas, it might make sense to curry the parameters. For example:
```python
def mk_pad_fn(length):
    return lambda field: str(field).strip().zfill(length)
# ...
data[...] = data[...].apply(mk_pad_fn(options['pad_to_length']))
```

This should also possible with `functools.partial`, though it feels less readable:
```python
data[...] = data[...].apply(partial(
    cls.padFunction, padTo=options['pad_to_length']))
```

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -25,85 +28,62 @@ def compareRecords (recordA,recordB, fields) :
+                # pad to specified length
+                data['{}'.format(col)] = data['{}'.format(col)].apply(
+                    lambda x: cls.padFunction(x, options['pad_to_length']))
+            if 'strip_commas' in options and options['strip_commas']:
```
Little thing, but this is slightly more Pythonic:
```python
if options.get('strip_commas'):
```

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -1,53 +1,216 @@
 
 def loadDomainValues(basePath, localSFPath = None, localProgramActivity = None):
-    """ Load all domain value files, localSFPath is used to point to a SF-133 file, if not provided it will be downloaded from S3  """
+    """Load all domain value files, localSFPath is used to point to a SF-133 file, if not provided it will be downloaded from S3."""
```
`localSFPath` no longer points to a file, but a directory. We should definitely update this docstring, but may also want to modify the variable name.

cmc333333 commented:
Added lots of little comments, but I think this is largely cleaner than what existed before. Unfortunately, tests are currently failing.

cmc333333 commented:
[run-tests]

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -130,30 +138,55 @@ def loadSF133(filename, fiscal_year, fiscal_period, force_load=False):
+                 'main_account_code', 'sub_account_code', 'tas', 'fiscal_year', 'period']
+    data.amount = data.amount.astype(float)  # this line triggers the settingwithcopy warning
+    data = pd.pivot_table(data, values='amount', index=pivot_idx, columns=['line'], fill_value=0).reset_index()
+    data = pd.melt(data, id_vars=pivot_idx, value_name='amount')
```
So many new terms to learn, like `melt`!

cmc333333 commented on dataactvalidator/filestreaming/loaderUtils.py:
```diff
@@ -65,7 +73,7 @@ def cleanData(cls, data, model, fieldMap, fieldOptions):
                 # pad to specified length
                 data['{}'.format(col)] = data['{}'.format(col)].apply(
-                    lambda x: cls.padFunction(x, options['pad_to_length']))
+                    cls.padFunction, args=(options['pad_to_length'],))
```
This threw me for a loop; I expected `args` to be the _first_ arguments to the function rather than the last. Good to know!

cmc333333 commented on dataactvalidator/filestreaming/loadFile.py:
```diff
@@ -199,9 +201,37 @@ def formatInternalTas(row):
+                prefix='{}/sf_133'.format(CONFIG_BROKER['sf_133_folder']))
+            sf133_list = [SF133File(sf133, os.path.basename(sf133.name)) for sf133 in sf133_files]
+
+    return sf133_list or []
```
There's a branch where the `sf133_list` variable isn't defined, but it may not be sensible:
`localSF133Dir is None` and `CONFIG_BROKER["use_aws"] == False`

cmc333333 replied:
Also, very nice refactor!

bsweger replied:
Thanks--will fix the branch thing. At some point, I would like to know more about what we should expect to happen for various permutations of the `use_aws` param and the local option, but will pull that thread another day.

bsweger commented:
@cmc333333 Okay, I made a lot of your suggested changes. Could keep tweaking, but I think the time has come for another look for things that need to get fixed before merge.

This is still WIP status while I check in w/ Katie for a heads-up on the one-time load of some historical data.

cmc333333 commented:
I really like the dataframe approach; I can't validate that business logic is identical but that's what tests are for :)
